xla/service/scatter_determinism_expander.cc

```diff
@@ -0,0 +1,459 @@
+HloComputation* ScalarSortingComparison(HloModule* module,
+                                        const Shape key_shape,
+                                        const Shape update_shape,
+                                        int64_t num_updates) {
+  HloComputation::Builder builder("sorting_computation");
+  auto param0 = builder.AddInstruction(
+      HloInstruction::CreateParameter(0, key_shape, "lhs_key"));
+  auto param1 = builder.AddInstruction(
+      HloInstruction::CreateParameter(1, key_shape, "rhs_key"));
+  const int kExistingParams = 2;
+  for (int i = 0; i < num_updates; ++i) {
+    builder.AddInstruction(
+        HloInstruction::CreateParameter(kExistingParams + i, update_shape,
+                                        absl::StrFormat("lhs_update_%d", i)));
+    builder.AddInstruction(
+        HloInstruction::CreateParameter(kExistingParams + 1 + i, update_shape,
+                                        absl::StrFormat("rhs_update_%d", i)));
+  }
+  builder.AddInstruction(
+      HloInstruction::CreateCompare(ShapeUtil::MakeShape(PRED, {}), param0,
+                                    param1, ComparisonDirection::kLt));
+  return module->AddEmbeddedComputation(builder.Build());
+}
+
+static std::vector<HloInstruction*> SortIndicesAndUpdates(
+    HloInstruction* scatter_indices,
+    const std::vector<HloInstruction*>& scatter_updates, int64_t num_indices,
+    HloScatterInstruction* scatter, HloComputation* parent) {
+  const Shape& indices_shape = scatter_indices->shape();
+  const Shape& updates_shape = scatter_updates[0]->shape();
+  auto updates_dims = updates_shape.dimensions();
+  // Since we canonicalized the scatter updates, the first dim will always be
+  // the number of updates and the rest will be the shape of each update
+
+  HloInstruction* scalar_indices = scatter_indices;
+
+  std::vector<int64_t> single_update_dimensions(updates_dims.begin() + 1,
+                                                updates_dims.end());
+
+  const Shape update_shape = ShapeUtil::MakeShape(updates_shape.element_type(),
+                                                  single_update_dimensions);
+
+  const Shape& scalar_index_shape =
+      ShapeUtil::MakeShape(indices_shape.element_type(), {num_indices});
+
+  auto* comparison = ScalarSortingComparison(
+      scatter->GetModule(),
+      ShapeUtil::MakeShape(indices_shape.element_type(), {}),
+      ShapeUtil::MakeShape(updates_shape.element_type(), {}),
+      scatter_updates.size());
+
+  std::vector<HloInstruction*> sort_operands = {scalar_indices};
+  std::vector<Shape> sort_shapes = {scalar_index_shape};
+  for (auto update : scatter_updates) {
+    sort_operands.push_back(update);
+    sort_shapes.push_back(update->shape());
+  }
+
+  auto* sorting = parent->AddInstruction(HloInstruction::CreateSort(
+      ShapeUtil::MakeTupleShape(sort_shapes), 0, sort_operands, comparison,
+      false /*is_stable*/));
```

Comment:
nit: we prefer annotations like /*is_stable=*/false

Reply:
changed